### PR TITLE
add hybrid nn system [wip]

### PR DESCRIPTION
**Pull request by christophmluscher**
add a hybrid nn system

still work in progress. very untested!!! but general feedback and input welcome

features are still missing but will be extended (seq disc training and restoring for example)

### CODE REVIEW

File: common/setups/rasr/gmm_system.py [hunk 1]
```diff
@@ -826,6 +830,46 @@ def sat_recognition(
...
+        gmm_output.lexicon = self.crp[corpus_key].lexicon_config
+        gmm_output.cart_tree = self.crp[
+            corpus_key
+        ].acoustic_model_config.state_tying.file
+        gmm_output.feature_flows = self.feature_flows[corpus_key]
+        gmm_output.features = self.feature_caches[corpus_key]
+        gmm_output.segment_path = self.crp[corpus_key].segment_path
+        gmm_output.allophone_file = self.allophone_files["base"]
+
+        if corpus_type == "dev" or corpus_type == "test":
```
Comment (michelwi) [inline common/setups/rasr/gmm_system.py:859]: Why only set LM for test sets? E.g. for sequence training we also have a LM for the train set.

File: common/setups/rasr/gmm_system.py [hunk 2]
```diff
@@ -826,6 +830,46 @@ def sat_recognition(
...
+            gmm_output.lm_file = self.crp[corpus_key].language_model_config.file
+            gmm_output.lm = self.crp[corpus_key].language_model_config
+            gmm_output.acoustic_mixtures = self.mixtures[corpus_key][
+                f"train_{steps.get_step_names_as_list()[step_idx - 1]}"
+            ][-1]
```
Comment (michelwi) [inline common/setups/rasr/gmm_system.py:864]: It always used to be the case, that the feature scorer is associated to a dev/test set. This makes sense, as there might be different settigns per test set.
But for the mixtures themselves it would make more sense to me to associate them to the respective train set they were trained on.

Actually, I think `self.mixtures[corpus_key][name]` only exists for train corpora anyway.

File: common/setups/rasr/gmm_system.py [hunk 3]
```diff
@@ -826,6 +830,46 @@ def sat_recognition(
...
+
+        if corpus_type == "train":
+            gmm_output.alignments = self.alignments[corpus_key][
+                f"train_{steps.get_step_names_as_list()[step_idx - 1]}"
+            ][-1]
```
Comment (michelwi) [inline common/setups/rasr/gmm_system.py:869]: Also dev sets can have alignments, e.g. if we use them for cross validation of a ce loss.

File: common/setups/rasr/gmm_system.py [hunk 4]
```diff
@@ -851,11 +895,17 @@ def run_monophone_step(self, step_args):
                 name = step_args.training_args["name"]
                 feature_scorer = (trn_c, f"train_{name}")
 
+                recog_args = copy.deepcopy(step_args.recognition_args)
+                if step_args.test_recognition_args is None:
+                    break
+                recog_args.update(step_args.test_recognition_args)
```
Comment (michelwi) [inline common/setups/rasr/gmm_system.py:901]: Maybe we do want to use the same recognition args for test and dev corpora?
```suggestion
                if step_args.test_recognition_args is not None:
                    recog_args.update(step_args.test_recognition_args)
```
or would you rather that
`test_recognition_args = None` -> no test
`test_recognition_args = {}` -> test with same config

File: common/setups/rasr/gmm_system.py [hunk 5]
```diff
@@ -887,11 +937,17 @@ def run_triphone_step(self, step_args):
                 name = step_args.training_args["name"]
                 feature_scorer = (trn_c, f"train_{name}")
 
+                recog_args = copy.deepcopy(step_args.recognition_args)
+                if step_args.test_recognition_args is None:
```
Comment (michelwi) [inline common/setups/rasr/gmm_system.py:941]: see above

File: common/setups/rasr/gmm_system.py [hunk 6]
```diff
@@ -940,7 +996,21 @@ def run_vtln_step(self, step_args, step_idx, steps):
                 )
 
             for tst_c in self.test_corpora:
-                pass
+                name = step_args.training_args["train"]["name"]
+                feature_scorer = (trn_c, f"train_{name}")
+
+                recog_args = copy.deepcopy(step_args.recognition_args)
+                if step_args.test_recognition_args is None:
+                    break
```
Comment (michelwi) [inline common/setups/rasr/gmm_system.py:1004]: see above

File: common/setups/rasr/gmm_system.py [hunk 7]
```diff
@@ -969,7 +1039,21 @@ def run_sat_step(self, step_args):
                 )
 
             for tst_c in self.test_corpora:
-                pass
+                name = step_args.training_args["name"]
+                feature_scorer = (trn_c, f"train_{name}")
+
+                recog_args = copy.deepcopy(step_args.recognition_args)
+                if step_args.test_recognition_args is None:
+                    break
```
Comment (michelwi) [inline common/setups/rasr/gmm_system.py:1047]: above

File: common/setups/rasr/gmm_system.py [hunk 8]
```diff
@@ -998,7 +1082,21 @@ def run_vtln_sat_step(self, step_args):
...
```
Comment (michelwi) [inline common/setups/rasr/gmm_system.py:1090]: a

File: common/setups/rasr/gmm_system.py [hunk 9]
```diff
@@ -1007,6 +1105,37 @@ def run_vtln_sat_step(self, step_args):
                     **step_args.sdm_args,
                 )
 
+    def run_recognition_step(self, step_args):
```
Comment (michelwi) [inline common/setups/rasr/gmm_system.py:1108]: Ive been writing "see above" now so many times that I would suggest to remove code duplication by moving out the two loops and calls to `self.recognition` into a separate function. Or maybe this `run_recognition_step` is already sufficient and you just call this in each training step.

File: common/setups/rasr/gmm_system.py [hunk 10]
```diff
@@ -998,7 +1082,21 @@ def run_vtln_sat_step(self, step_args):
...
+                name = step_args.training_args["name"]
+                feature_scorer = (trn_c, f"train_{name}")
+
+                recog_args = copy.deepcopy(step_args.recognition_args)
+                if step_args.test_recognition_args is None:
+                    break
+                recog_args.update(step_args.test_recognition_args)
+                recog_args["optimize_am_lm_scale"] = False
+
+                self.recognition(
```
Comment (michelwi) [inline common/setups/rasr/gmm_system.py:1094]:
```suggestion
                self.sat_recognition(
```

File: common/setups/rasr/gmm_system.py [hunk 11]
```diff
@@ -969,7 +1039,21 @@ def run_sat_step(self, step_args):
...
```
Comment (michelwi) [inline common/setups/rasr/gmm_system.py:1051]:
```suggestion
                self.sat_recognition(
```

File: common/setups/rasr/util.py [hunk 12]
```diff
@@ -244,10 +251,15 @@ def __init__(
             }
         }
         ##################################################
+        test_recognition_args
```
Comment (michelwi) [inline common/setups/rasr/util.py:254]: If you keep it this way, add comment that the default `None` leads to no test recognition
```suggestion
        :param test_recognition_args:
```

File: common/setups/rasr/util.py [hunk 13]
```diff
@@ -382,9 +407,14 @@ def __init__(
         ##################################################
         :param recognition_args:
         ##################################################
+        test_recognition_args
```
Comment (michelwi) [inline common/setups/rasr/util.py:410]:
```suggestion
        :param test_recognition_args:
```

File: common/setups/rasr/util.py [hunk 14]
```diff
@@ -349,6 +368,10 @@ def __init__(
...
```
Comment (michelwi) [inline common/setups/rasr/util.py:371]:
```suggestion
        :param test_recognition_args:
```

File: common/setups/rasr/util.py [hunk 15]
```diff
@@ -309,6 +322,10 @@ def __init__(
...
```
Comment (michelwi) [inline common/setups/rasr/util.py:325]:
```suggestion
        :param test_recognition_args:
```

File: common/setups/rasr/util.py [hunk 16]
```diff
@@ -429,47 +461,235 @@ def __init__(
...
+        self.recognition_args = recognition_args
+
+
+class OutputArgs:
+    def __init__(self, name):
+        self.name = name
+        self.corpus_key_type_pair = {}
+
+    def add_corpus_key_type_pair(self, corpus_key, corpus_type):
+        self.corpus_key_type_pair[corpus_key] = corpus_type
```
Comment (michelwi) [inline common/setups/rasr/util.py:476]: optional cosmetic change:
```suggestion
        self.corpus_type_mapping = {}

    def define_corpus_type(self, corpus_key, corpus_type):
        self.corpus_type_mapping[corpus_key] = corpus_type
```

**Review by michelwi — COMMENTED** [review]
some comments on everything but the new NN system.

File: common/setups/rasr/gmm_system.py (hunk 1 above)
Comment (christophmluscher) [reply common/setups/rasr/gmm_system.py:859]: Yes. But this is the GMM output. I thought that for sequence training we normally want a different LM for lattice generation. For example, Librispeech count LM for decoding is 4gram. For sequence training, this is a 2gram. or not?
That is why I did not want to automatically set the LM for training.. but I am open to handle it differently

File: common/setups/rasr/gmm_system.py (hunk 3 above)
Comment (christophmluscher) [reply common/setups/rasr/gmm_system.py:869]: I agree. I ll add this

File: common/setups/rasr/gmm_system.py (hunk 4 above)
Comment (christophmluscher) [reply common/setups/rasr/gmm_system.py:901]: yes. I want to have no decoding for the `None` case.
Idea is to first evaluate on the `dev` set. Find a good set of parameters and then pick/update the recog args accordingly

File: common/setups/rasr/gmm_system.py (hunk 1 above)
Comment (michelwi) [reply common/setups/rasr/gmm_system.py:857]: Yes, so for train set a 2gram LM would be exported.

This is not about automatically setting it, but about exporting it if set.

File: common/setups/rasr/util.py [hunk 17]
```diff
@@ -8,8 +8,13 @@
     "GmmSatArgs",
     "GmmVtlnSatArgs",
     "ForcedAlignmentArgs",
+    "RecognitionArgs",
```
Comment (JackTemaki) [inline common/setups/rasr/util.py:11]: Do you think it would make sense to separate this further? I imagine that GmmRecogntionArgs, HybridRecognitionArgs, CTCRegognitionArgs ect.... might be difficult.

File: common/setups/rasr/util.py [hunk 18]
```diff
@@ -429,47 +470,239 @@ def __init__(
...
+    ):
+        self.oggzip_files = oggzip_files
+        self.alignments = alignments
+        self.context_window = (
+            context_window
+            if context_window is not None
+            else {"classes": 1, "data": 242}
+        )
+        self.audio = (
+            audio if audio is not None else {"features": "raw", "sample_rate": 16000}
```
Comment (JackTemaki) [inline common/setups/rasr/util.py:641]: I think it is not good to have hardcoded values somehwere here.

File: common/setups/rasr/util.py [hunk 19]
```diff
@@ -429,47 +470,239 @@ def __init__(
...
+
+class OggZipHdfDataInput:
+    def __init__(
+        self,
+        oggzip_files: tk.Path,
+        alignments: tk.Path,
+        context_window: Optional[Dict] = None,
+        audio: Optional[Dict] = None,
+        partition_epoch: int = 1,
+        seq_ordering: str = "laplace:226",
```
Comment (JackTemaki) [inline common/setups/rasr/util.py:630]: this is a bad default, sequence ordering should be data size independent, so something like `laplace:.1000` (the dot says that this should be the bin size instead of the number of bins)

File: common/setups/rasr/util.py [hunk 20]
```diff
@@ -429,47 +470,239 @@ def __init__(
...
+        targets: Optional[str] = None,
+    ):
+        self.oggzip_files = oggzip_files
+        self.alignments = alignments
+        self.context_window = (
+            context_window
+            if context_window is not None
+            else {"classes": 1, "data": 242}
```
Comment (JackTemaki) [inline common/setups/rasr/util.py:638]: I do not understand what this is about.

File: common/setups/rasr/util.py [hunk 21]
```diff
@@ -429,47 +470,239 @@ def __init__(
...
+            crp.language_model_config.file = self.lm["filename"]
+            crp.language_model_config.scale = self.lm["scale"]
+
+        if self.allophone_file is not None:
+            crp.acoustic_model_config.allophones.add_from_file = self.allophone_file
+
+        return crp
+
+
+class OggZipHdfDataInput:
```
Comment (JackTemaki) [inline common/setups/rasr/util.py:622]: Can you already add docstrings to the new stuff? Then it would be much easier for me to understand why you are actually introducing this, and then it is also already better documented.

File: common/setups/rasr/util.py [hunk 22]
```diff
@@ -309,6 +323,11 @@ def __init__(
         ##################################################
         :param recognition_args:
         ##################################################
+        :param test_recognition_args:
+        decoding parameters might change depending on whether dev or test sets are decoded.
+        setting test_recognition_args to None means to test recognition!
```
Comment (JackTemaki) [inline common/setups/rasr/util.py:328]: to -> no?

**Review by JackTemaki — CHANGES_REQUESTED** [review]
Maybe it is a good idea to split "Util" a little bit into the parts for which is actually needed. But this can be discussed.

File: common/setups/rasr/util.py (hunk 17 above)
Comment (christophmluscher) [reply common/setups/rasr/util.py:11]: I am thinking that maybe we could merge the `Gmm*Args` to `GmmStepArgs` because basically it is always the same steps and then make an documentation at the beginning of the file.

at the moment RecognitionArgs only holds the name plus a dict with the arguments.. I dont see a need to split this..

I generally agree that the `util.py` is a bit big

File: common/setups/rasr/util.py (hunk 20 above)
Comment (christophmluscher) [reply common/setups/rasr/util.py:638]: the values for the OggZipHdfData are taken from a setup by Peter. So for raw waveform setups. I can remove the defaults for `audio` and `context_window`

Comment (vieting) [reply common/setups/rasr/util.py:638]: Yes, I think this default value should be removed.

File: common/setups/rasr/util.py (hunk 18 above)
Comment (vieting) [reply common/setups/rasr/util.py:641]: This one is more likely to be helpful but still, I agree we shouldn't hardcode it here.

File: common/setups/rasr/util.py (hunk 21 above)
Comment (christophmluscher) [reply common/setups/rasr/util.py:622]: yes I can

File: common/setups/rasr/util.py (hunk 17 above)
Comment (christophmluscher) [reply common/setups/rasr/util.py:11]: split maybe into `gmm_util`, `nn_util`, `ctc_util` or so?

File: common/setups/rasr/util.py [hunk 23]
```diff
@@ -429,47 +470,242 @@ def __init__(
...
+        audio: Dict,
+        targets: str,
+        partition_epoch: int = 1,
+        seq_ordering: str = "laplace:.1000",
+    ):
+        """
+        :param oggzip_files:
+        :param alignments:
+        :param context_window: {"classes": 1, "data": 242}
+        :param audio: {"features": "raw", "sample_rate": 16000}
```
Comment (vieting) [inline common/setups/rasr/util.py:637]: Maybe explain what this would mean
```suggestion
        :param audio: e.g. {"features": "raw", "sample_rate": 16000} for raw waveform input with a sample rate of 16 kHz
```

File: common/setups/rasr/nn_system.py [hunk 24]
```diff
@@ -0,0 +1,554 @@
...
+        if "cropped" in network:
+            if network["output"]["from"] == ["cropped"]:
+                network["output"]["from"] = "upsample"
+            network.pop("cropped")
+            if "lstm_bwd_1" in network:
+                network["lstm_bwd_1"]["from"] = "upsample"
+                network["lstm_fwd_1"]["from"] = "upsample"
+            if "lstm_fwd_1_no_init" in network:
+                network["lstm_bwd_1_no_init"]["from"] = "upsample"
+                network["lstm_fwd_1_no_init"]["from"] = "upsample"
```
Comment (michelwi) [inline common/setups/rasr/nn_system.py:136]: this is ugly special handling with hardcoded layer names.. 

maybe rather allow for arbitrary user definable cleanup functions that are called on the network..

File: common/setups/rasr/nn_system.py [hunk 25]
```diff
@@ -0,0 +1,554 @@
...
+
+        self._init_am(**self.hybrid_init_args.am_args)
+
+        devtrain_data = devtrain_data if devtrain_data is not None else {}
+        dev_data = dev_data if dev_data is not None else {}
+        test_data = test_data if test_data is not None else {}
+
+        self._assert_corpus_name_unique(
+            train_data, devtrain_data, cv_data, dev_data, test_data
+        )
```
Comment (michelwi) [inline common/setups/rasr/nn_system.py:248]: does this mean cv_data == dev_data is not supportet?

Comment (michelwi) [reply common/setups/rasr/nn_system.py:248]: Also reordering for consistency reasons
```suggestion
        self._assert_corpus_name_unique(
            train_data, cv_data, devtrain_data, dev_data, test_data
        )
```

File: common/setups/rasr/nn_system.py [hunk 26]
```diff
@@ -0,0 +1,554 @@
...
+                raw_sample_rate=raw_sampling_rate,
+                feat_sample_rate=feature_sampling_rate,
+            )
+            self.oggzips[name] = j.out_ogg_zip
+            j.add_alias(f"oggzip/{name}")
+
+            # TODO self.jobs[name]["hdf_full"] = j = returnn.ReturnnDumpHDFJob()
+
+    def generate_lattices(self):
+        pass
```
Comment (michelwi) [inline common/setups/rasr/nn_system.py:287]: technically we could even do sequence training for GMMs. Just nobody since Zoltan has done this and I don't even have an example config for this.

So while in general this function could go into the `RasrSystem`, we can also leave it here.

Comment (michelwi) [reply common/setups/rasr/nn_system.py:287]: I can do a PR to add the lattices if you want. Just remind me some time ;)

File: common/setups/rasr/nn_system.py [hunk 27]
```diff
@@ -0,0 +1,554 @@
...
+        self.nn_models[f"{train_corpus_key}_{cv_corpus_key}"][
+            name
+        ] = train_job.out_models
+        self.nn_configs[f"{train_corpus_key}_{cv_corpus_key}"][
+            name
+        ] = train_job.out_returnn_config_file
+
+        return train_job
+
+    def returnn_rasr_training(
```
Comment (michelwi) [inline common/setups/rasr/nn_system.py:329]: This function shares almost the same signature as `returnn_training` and does very similar things. 
Would it be possible / desirable to only have one function, which then internally decides on `ReturnnTrainingJob` vs `ReturnnRasrTrainingJob` and does the handling based on `*_corpus_key`.

File: common/setups/rasr/nn_system.py [hunk 28]
```diff
@@ -0,0 +1,554 @@
...
+        self,
+        name: str,
+        returnn_config: returnn.ReturnnConfig,
+        checkpoints: List[returnn.Checkpoint],
+        acoustic_mixture_path: tk.Path,
+        prior_scale: float,
+        pronunciation_scales: List[float],
+        lm_scales: List[float],
+        optimize_am_lm_scale: bool,
+        corpus_key: str,
```
Comment (michelwi) [inline common/setups/rasr/nn_system.py:380]:
```suggestion
        recognition_corpus_key: str,
```
?

File: common/setups/rasr/nn_system.py [hunk 29]
```diff
@@ -0,0 +1,554 @@
...
+        return train_job
+
+    # -------------------- Recognition --------------------
+
+    def nn_recognition(
+        self,
+        name: str,
+        returnn_config: returnn.ReturnnConfig,
+        checkpoints: List[returnn.Checkpoint],
+        acoustic_mixture_path: tk.Path,
```
Comment (michelwi) [inline common/setups/rasr/nn_system.py:375]: `acoustic_mixture_path` could be optional, if a prior file is provided.
(ok, not really.. we would still need a dummy mixture)

File: common/setups/rasr/nn_system.py [hunk 30]
```diff
@@ -0,0 +1,554 @@
...
+        prior_scale: float,
```
Comment (michelwi) [inline common/setups/rasr/nn_system.py:376]: mh.. maybe `prior_scales` also deserves to be a list to be iterated over?

File: common/setups/rasr/nn_system.py [hunk 31]
```diff
@@ -0,0 +1,554 @@
...
+
+    def run_nn_step(self, step_args):
+        for trn_c, cv_c, dvtr_c, name in itertools.product(
+            self.train_corpora,
+            self.cv_corpora,
+            self.devtrain_corpora,
+            step_args.returnn_configs,
+        ):
+            if (trn_c, cv_c) in self.train_cv_pairing or self.train_cv_pairing is None:
+                continue
```
Comment (michelwi) [inline common/setups/rasr/nn_system.py:455]: do you really mean to skip over all combinations defined in the pairing or is there a `not` missing?

Comment (michelwi) [reply common/setups/rasr/nn_system.py:455]: I am wondering if iterating over `itertools.product(..)` is the right thing to do, or if in the step args a list of `(trn_c, cv_c, dvtr_c, name)` tuples can be defined that is iterated over.

**Review by michelwi — COMMENTED** [review]
some comments on `rasr/nn_system.py`

File: common/setups/rasr/nn_system.py [hunk 32]
```diff
@@ -0,0 +1,554 @@
...
+        self.datasets = {}
+
+        self.oggzips = {}
+
+        self.nn_configs = {}
+        self.nn_models = {}
+
+    # -------------------- Helpers --------------------
+    @staticmethod
+    def adapt_returnn_config_for_recog(returnn_config: returnn.ReturnnConfig):
```
Comment (JackTemaki) [inline common/setups/rasr/nn_system.py:105]: I think if this function is needed at all there are severe flaws in how we are building our configs. I am trying to solve the same issues right now for the CTC setup. I will give you am update when I have a clean solution for this.

File: common/setups/rasr/nn_system.py [hunk 33]
```diff
@@ -0,0 +1,554 @@
...
+            if "lstm_bwd_1" in network:
+                network["lstm_bwd_1"]["from"] = "upsample"
+                network["lstm_fwd_1"]["from"] = "upsample"
+            if "lstm_fwd_1_no_init" in network:
+                network["lstm_bwd_1_no_init"]["from"] = "upsample"
+                network["lstm_fwd_1_no_init"]["from"] = "upsample"
+
+        return config
+
+    @staticmethod
```
Comment (JackTemaki) [inline common/setups/rasr/nn_system.py:140]: I think this helper can be external at some point. Or actually I see a lot of overlap potential in my current CTCSystem vs the NNSystem here. I think it would be nice if the CTCSystem can fully extend from it. Then this would be fine here, but still people might want to build flows without the System class. All other flows are also defined standalone. (But there is also no problem in moving this later)

File: common/setups/rasr/nn_system.py (hunk 25 above)
Comment (JackTemaki) [reply common/setups/rasr/nn_system.py:248]: > does this mean cv_data == dev_data is not supportet?

Unfortunately this is still impossible with the current Training jobs :/

File: common/setups/rasr/nn_system.py [hunk 34]
```diff
@@ -0,0 +1,554 @@
...
+        self,
+        returnn_root: Optional[str] = None,
+        returnn_python_home: Optional[str] = None,
+        returnn_python_exe: Optional[str] = None,
+    ):
+        super().__init__()
+
+        self.returnn_root = returnn_root or gs.RETURNN_ROOT
+        self.returnn_python_home = returnn_python_home or gs.RETURNN_PYTHON_HOME
+        self.return_python_exe = returnn_python_exe or gs.RETURNN_PYTHON_EXE
```
[the remark this answers is not in the document]
Comment (christophmluscher) [reply common/setups/rasr/nn_system.py:81]: My thinking was that `ReturnnTrainingJob` and `ReturnnRasrTrainingJob` also both require setting `RETURNN_*` in `gs`. Then I would rather have the setup fail as soon as possible.
But I did not think about setting `returnn_` at a later stage, e.g. via `step_args`.
I think you are right to leave this open so that the user has flexibility.

File: common/setups/rasr/nn_system.py [hunk 35]
```diff
@@ -0,0 +1,554 @@
...
+        self.devtrain_corpora = []
+        self.cv_corpora = []
+
+        self.train_input_data = None
+        self.cv_input_data = None
+        self.devtrain_input_data = None
+        self.dev_input_data = None
+        self.test_input_data = None
+
+        self.train_cv_pairing = None
```
[the remark this answers is not in the document]
Comment (christophmluscher) [reply common/setups/rasr/nn_system.py:94]: At the moment, you can add two train datasets and two cv datasets. This would result in 4 training jobs. If you want to fix train/cv pairs, this variable holds this information.

File: common/setups/rasr/nn_system.py (hunk 34 above)
[the remark this answers is not in the document]
Comment (michelwi) [reply common/setups/rasr/nn_system.py:81]: There is no harm in having it fail here if it is unset. But then rather than throwing an `AttributeError` we should in this case assert that either is set.

File: common/setups/rasr/nn_system.py [hunk 36]
```diff
@@ -0,0 +1,554 @@
...
+
+        self.datasets = {}
+
+        self.oggzips = {}
```
[the remark this answers is not in the document]
Comment (christophmluscher) [reply common/setups/rasr/nn_system.py:98]: I am not sure how to handle this yet. WIP. At some point, I would like to have the possibility to automatically convert a dataset based on CRPs to a setup with OggZips and/or ... I think that would need the extra variables, as you said, but not fully thought through yet.

[the remark this answers is not in the document]
Comment (christophmluscher) [reply common/setups/rasr/nn_system.py:98]: One idea: use `datasets` for `ReturnnRasrTrainingJob`. Everything else for `ReturnnTrainingJob`. Or so

File: common/setups/rasr/nn_system.py [hunk 37]
```diff
@@ -0,0 +1,554 @@
...
+        Adapt a RETURNN config for recognition, e.g., remove loss and use log softmax activation in last layer
+
+        :param ReturnnConfig config:
+        :rtype ReturnnConfig:
+        """
+        assert isinstance(returnn_config, returnn.ReturnnConfig)
+        config = copy.deepcopy(returnn_config)
+        forward_output_layer = config.config.get("forward_output_layer", "output")
+        network = config.config.get("network")
+        for layer_name, layer in network.items():
```
[the remark this answers is not in the document]
Comment (christophmluscher) [reply common/setups/rasr/nn_system.py:116]: I agree. Based on comments by Nick. I ll wait to see his solution which should hopefully mean that we can remove this function

File: common/setups/rasr/nn_system.py [hunk 38]
```diff
@@ -0,0 +1,554 @@
...
+            if layer.get("unit", None) in {"lstmp"}:
+                layer["unit"] = "nativelstm2"
+            if layer.get("target", None) and layer_name != forward_output_layer:
+                layer.pop("target")
+                layer.pop("loss", None)
+                layer.pop("loss_scale", None)
+        network[forward_output_layer]["class"] = "linear"
+        network[forward_output_layer]["activation"] = "log_softmax"
+        network[forward_output_layer].pop("target", None)
```
[the remark this answers is not in the document]
Comment (christophmluscher) [reply common/setups/rasr/nn_system.py:125]: I dont know. Nick hopefully makes this function obsolete :)

[the remark this answers is not in the document]
Comment (christophmluscher) [reply common/setups/rasr/nn_system.py:125]: @JackTemaki no pressure :D

File: common/setups/rasr/nn_system.py (hunk 24 above)
Comment (christophmluscher) [reply common/setups/rasr/nn_system.py:136]: true

File: common/setups/rasr/nn_system.py (hunk 26 above)
Comment (christophmluscher) [reply common/setups/rasr/nn_system.py:287]: Ah then I would prefer to leave this function here :)

> I can do a PR to add the lattices if you want. Just remind me some time ;)

yes please :)

File: common/setups/rasr/nn_system.py (hunk 27 above)
Comment (christophmluscher) [reply common/setups/rasr/nn_system.py:329]: maybe not sure :)

File: common/setups/rasr/nn_system.py (hunk 31 above)
Comment (christophmluscher) [reply common/setups/rasr/nn_system.py:455]: yes. good point. that could be nicer and maybe more user-friendly..

File: common/setups/rasr/nn_system.py (hunk 33 above)
Comment (christophmluscher) [reply common/setups/rasr/nn_system.py:140]: Yes, we should maybe discuss this on Tuesday, how CTCSystem and NNSystem are going to relate to each other.

maybe a `flow` dir in `setups`??

Comment (JackTemaki) [reply common/setups/rasr/nn_system.py:146]: maybe this could be in i6_core/returnn/flow.py or so...? I am not sure.

File: common/setups/rasr/nn_system.py [hunk 39]
```diff
@@ -0,0 +1,564 @@
...
+        self.test_input_data = None
+
+        self.train_cv_pairing = None
+
+        self.datasets = {}
+
+        self.oggzips = {}
+
+        self.nn_configs = {}
+        self.nn_models = {}
```
Comment (JackTemaki) [inline common/setups/rasr/nn_system.py:97]: I just saw that there is a big inconsistency with respect to "Models" and "Checkpoints". We should only use one, so that all stuff like "getting best epoch", "checkpoint averaging" and so on will be working. I think you switched to `Checkpoint` in your further code, and I would suggest to do this everywhere, also here.

File: common/setups/rasr/nn_system.py [hunk 40]
```diff
@@ -0,0 +1,564 @@
...
+        )
+
+        tf_flow.config[tf_fwd].output_map.info_0.param_name = "log-posteriors"
+        tf_flow.config[
+            tf_fwd
+        ].output_map.info_0.tensor_name = f"{forward_output_layer}/output_batch_major"
+
+        tf_flow.config[tf_fwd].loader.type = "meta"
+        tf_flow.config[tf_fwd].loader.meta_graph_file = tf_graph_path
+        tf_flow.config[tf_fwd].loader.saved_model_file = checkpoint_path
```
Comment (JackTemaki) [inline common/setups/rasr/nn_system.py:193]: Are you sure this worked for you? I got some problems when directly passing the checkpoint, as the path had the `*.index` extension which RASR did not accept, so I needed to add a "delayed cutting" function to solve this.

Comment (michelwi) [reply common/setups/rasr/nn_system.py:193]: I think this was the problem with `Model` and we specifically designed `Checkpoint` such that this works..
Here (https://github.com/rwth-i6/i6_core/blob/deb2bbe5748e9f9e4e9c6740290928beb5f1edf9/returnn/training.py#L55) the additional `".index"` should be removed.

Comment (michelwi) [reply common/setups/rasr/nn_system.py:193]: >  I will set `ReturnnModel` to deprecated

I don't think I meant `ReturnnModel` (which contains a path, a config, and an epoch) but a different class that was effectively doing the same as `Checkpoint` but was hashing the absolute path to the model. (And was already removed)

> The `ckpt_path` can not be used here, because this already calls get_path()

But we would only set the `Checkpoint` object here. The call to `Checkpoint.str` (and subsequently `ckpt_path`) would only happen at runtime when the `create_files` Task of Recognition was executed; Therefore we _can_ use it here, right?

File: common/setups/rasr/nn_system.py (hunk 37 above)
[the remark this answers is not in the document]
Comment (JackTemaki) [reply common/setups/rasr/nn_system.py:116]: Okay, so we can not completely drop this function I think, but there are some things that can be removed.

 - Targets do not to be dropped in RETURNN, for me I can always keep them.
 - I would consider using lstmp an user error, and not something to fix here
 - For the log outputs, they can be part of the network in the first place, and then we can tell the tf_flow helper what output to use. This would be more consistent and flexible (and safe heuristics code).
 
 For the "cropped" thing I would refer to @michelwi solution, as this is way too specific to be included here.

File: common/setups/rasr/nn_system.py [hunk 41]
```diff
@@ -454,7 +454,7 @@ def nn_recognition(
         self,
         name: str,
         returnn_config: returnn.ReturnnConfig,
-        checkpoints: List[returnn.Checkpoint],
+        returnn_models: List[returnn.ReturnnModel],
```
Comment (JackTemaki) [inline common/setups/rasr/nn_system.py]: Why did you change the Checkpoints back to Model? Is there something the Checkpoint can not do? Because I remember that we discussed at some point that ReturnnModel should be deprecated because it is not supported by some jobs (e.g. CheckpointAveraging or GetBestCheckpoint).

Comment (christophmluscher) [reply common/setups/rasr/nn_system.py]: No, I wrote that down incorrectly. I ll revert back then

Comment (JackTemaki) [reply common/setups/rasr/nn_system.py]: But why was that incorrectly? I am using `Checkpoint` for my CTCSystem and this seems to be more consistent with the rest of the setups.

Comment (christophmluscher) [reply common/setups/rasr/nn_system.py]: I wrote down that Checkpoint was deprecated. I didn't recheck that. My mistake

Comment (christophmluscher) [reply common/setups/rasr/nn_system.py]: That was incorrect of me

**Review by JackTemaki — APPROVED** [review]
Lets merge this and then continue!

**Comment by JackTemaki** [comment]
@christophmluscher you need to rebase the branch once on the master, there are merge conflicts!

**Comment by michelwi** [comment]
> Lets merge this and then continue!

There are many changes in user folders of other users (rossenbach/dierkes). Is this intended or a rebase-bug?

**Comment by christophmluscher** [comment]
> you need to rebase the branch once on the master, there are merge conflicts!

yes. will do

> There are many changes in user folders of other users (rossenbach/dierkes). Is this intended or a rebase-bug?

I did a merge from main, that was maybe not optimal. I ll look into this
